### PR TITLE
[API Compatibility] add aminmax op-part

### PR DESCRIPTION
**Pull request by Manfredss**
### PR Category
User Experience

### PR Types
New features

### Description
- Add `paddle.aminmax` API that returns both minimum and maximum of tensor elements over the given axis, aligned with
`torch.aminmax`
- Supports `axis`, `keepdim` parameters and PyTorch-compatible aliases (`input` for `x`, `dim` for `axis`)
- Backward gradient correctly distributes evenly among duplicate min/max elements (same as `amin`/`amax`)

### New files
- `paddle/phi/kernels/aminmax_kernel.h/.cc` — forward kernel (reuses `AMinRawKernel` + `AMaxRawKernel`)
- `paddle/phi/kernels/aminmax_grad_kernel.h/.cc` — backward kernel (`amax_grad + amin_grad`)
- `test/legacy_test/test_aminmax_op.py` — unit tests

### Modified files
- `paddle/phi/ops/yaml/ops.yaml` / `backward.yaml` — op and grad definitions
- `paddle/phi/ops/yaml/python_api_info.yaml` / `op_compat.yaml` — API info and compat
- `paddle/phi/infermeta/unary.h/.cc` — `AMinMaxInferMeta` (two-output reduce shape)
- `.../infer_symbolic_shape/unary_infer_sym.h/.cc` — `AminmaxOpInferSymbolicShape`
- `python/paddle/__init__.py`, `python/paddle/tensor/__init__.py` — exports
- `python/paddle/_paddle_docs.py` — docstring and signature
- `test/legacy_test/test_api_compatibility.py` — API compatibility test

### API Signature
```python
paddle.aminmax(x, axis=None, keepdim=False, name=None) -> tuple[Tensor, Tensor]
```

### Test plan
- test_aminmax_op.py — 27 tests (axis variants, keepdim, dtypes, zero-dim, gradient)
- test_api_compatibility.py::TestAminmaxAPI — dygraph + static, paddle/torch keyword aliases
- Passed all 15 tests on enhanced PaConvert tests

### 是否引起精度变化
否

### CODE REVIEW

File: python/paddle/_paddle_docs.py [hunk 1]
```diff
@@ -320,6 +320,76 @@ def amin(
...
+    "aminmax",
+    r"""
+    Computes both the minimum and maximum of tensor elements over the given axis.
+
+    Note:
+        Like amin and amax, if there are multiple minimum/maximum elements,
+        aminmax evenly distributes gradient between these equal values.
+
+    Args:
+        x (Tensor): A tensor, the data type is float32, float64, int32, int64.
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py:333]: Alias写下

File: paddle/phi/ops/yaml/op_compat.yaml [hunk 2]
```diff
@@ -206,6 +206,18 @@
     amin_grad : GetReduceGradExpectedKernelType
   manual_signature : [amin]
 
+- op : aminmax
```
Comment (zhwesky2010) [inline paddle/phi/ops/yaml/op_compat.yaml:209]: 这个应该是一些不兼容历史问题的适配，这个需要改吗，不报错可以不改

File: python/paddle/_paddle_docs.py [hunk 3]
```diff
@@ -320,6 +320,76 @@ def amin(
...
+        axis (int|list|tuple|None, optional): The axis along which the minimum and maximum
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py:335]: Alias

File: python/paddle/_paddle_docs.py [hunk 4]
```diff
@@ -320,6 +320,76 @@ def amin(
...
+            are computed. If :attr:`None`, compute over all elements of
+            `x` and return Tensors with a single element,
+            otherwise must be in the range :math:`[-x.ndim, x.ndim)`.
+            If :math:`axis[i] < 0`, the axis to reduce is :math:`x.ndim + axis[i]`.
+        keepdim (bool, optional): Whether to reserve the reduced dimension in the
+            output Tensors. The result tensors will have one fewer dimension
+            than the `x` unless :attr:`keepdim` is true, default
+            value is False.
+        name (str|None, optional): Name for the operation (optional, default is None). For more information, please refer to :ref:`api_guide_Name`.
+
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py:345]: 关键字参数文档

File: test/legacy_test/test_aminmax_op.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+import numpy as np
+
+import paddle
+
+
+class TestAminmaxAPI(unittest.TestCase):
+    def setUp(self):
+        self.init_case()
+        self.place = paddle.CPUPlace()
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py]: 无需设置place，新增OP还是在每个设备上都测下较好

File: test/legacy_test/test_aminmax_op.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+            )
+
+            np.testing.assert_allclose(fetches[0], expected_min, rtol=1e-05)
+            np.testing.assert_allclose(fetches[1], expected_max, rtol=1e-05)
+
+    def test_dygraph_gradient(self):
+        paddle.disable_static()
+        x = paddle.to_tensor(self.x_np, dtype=self.dtype, stop_gradient=False)
+        min_val, max_val = paddle.aminmax(x, self.axis, self.keepdim)
+        loss = min_val.sum() + max_val.sum()
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py:138]: 反向测下值吧

File: test/legacy_test/test_aminmax_op.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+# limitations under the License.
+
+import unittest
+
+import numpy as np
+
+import paddle
+
+
+class TestAminmaxAPI(unittest.TestCase):
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py]: 新增OP可以测全一些，可以继承OPTest基类来测下。

File: python/paddle/_paddle_docs.py [hunk 8]
```diff
@@ -320,6 +320,76 @@ def amin(
 """,
 )
 
+add_doc_and_signature(
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py:323]: 文档统一按这个格式规范来吧：

**英文-python装饰器**
https://github.com/PaddlePaddle/docs/blob/develop/docs/dev_guides/coding_agent/api_compatibility/python-decorator.md#step-4-%E6%9B%B4%E6%96%B0%E5%87%BD%E6%95%B0%E6%96%87%E6%A1%A3%E5%AD%97%E7%AC%A6%E4%B8%B2

**英文-cpp下沉**
https://github.com/PaddlePaddle/docs/blob/develop/docs/dev_guides/coding_agent/api_compatibility/cpp-sink.mdr#L66

**中文**
https://github.com/PaddlePaddle/docs/blob/develop/docs/dev_guides/coding_agent/api_compatibility/api-docs-updater.mdr

**Comment by zhwesky2010** [comment]
覆盖率没过
infoflow 2026-03-25 11-21-27

File: paddle/phi/ops/yaml/op_compat.yaml [hunk 9]
```diff
@@ -206,6 +206,16 @@
...
```
Comment (zhwesky2010) [inline paddle/phi/ops/yaml/op_compat.yaml:209]: 这个是否可以不改？

File: test/legacy_test/test_aminmax_op.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+        # 1. Paddle Positional arguments
+        min1, max1 = paddle.aminmax(x)
+        # 2. Paddle keyword arguments
+        min2, max2 = paddle.aminmax(x=x)
+        # 3. PyTorch keyword arguments (alias)
+        min3, max3 = paddle.aminmax(input=x)
+        # 4. Mixed arguments (positional x + keyword axis)
+        min4, max4 = paddle.aminmax(x, axis=0)
+        # 5. PyTorch keyword arguments (dim alias)
+        min5, max5 = paddle.aminmax(x, dim=0)
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py]: 测一下out参数

File: python/paddle/_paddle_docs.py [hunk 11]
```diff
@@ -320,6 +320,78 @@ def amin(
...
+            Tensor(shape=[2, 1], dtype=float64, place=Place(cpu), stop_gradient=False,
+            [[0.90000000],
+             [0.90000000]])
+""",
+    """
+def aminmax(
+    x: Tensor,
+    axis: int | Sequence[int] | None = None,
+    keepdim: bool = False,
+    name: str | None = None,
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py]: 默认是支持out为关键字参数的，所有op会默认解析kwargs: out

File: paddle/phi/ops/yaml/op_compat.yaml (hunk 9 above)
Comment (Manfredss) [reply paddle/phi/ops/yaml/op_compat.yaml:209]: 这里的定义用于映射旧框架大写命名（X、Min、Max）到新 phi 内核小写命名（x、min、max）。OpTest 类依赖此映射来构建算子，移除后 check_output 和 check_grad 会报 Missing x as input 或 out_dtype not found 错误，已通过验证不能去掉

File: paddle/phi/ops/yaml/op_compat.yaml [hunk 12]
```diff
@@ -206,6 +206,16 @@
...
+  backward : aminmax_grad
+  inputs :
+    x : X
+  outputs :
+    {min : Min, max : Max}
+  extra :
+    attrs : [bool use_mkldnn = false, bool use_onednn = false]
```
Comment (zhwesky2010) [inline paddle/phi/ops/yaml/op_compat.yaml]: 这些extra应该是没有的

File: python/paddle/_paddle_docs.py [hunk 13]
```diff
@@ -320,6 +320,83 @@ def amin(
...
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py]: 新增API统一不支持name

File: test/legacy_test/test_aminmax_op.py [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+
+
+class TestAminmaxOpFloat32(TestAminmaxOp):
+    def init_dtype(self):
+        self.dtype = np.float32
+
+    def init_shape(self):
+        self.shape = [10, 10]
+
+    def test_check_grad(self):
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py:118]: 这个是什么问题

File: test/legacy_test/test_aminmax_op.py [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+class TestAminmaxCompatibility(unittest.TestCase):
+    """Test paddle/torch keyword alias compatibility."""
+
+    def setUp(self):
+        np.random.seed(2025)
+        self.np_x = np.array(
+            [[0.2, 0.3, 0.5, 0.9], [0.1, 0.2, 0.6, 0.7]], dtype='float64'
+        )
+
+    def test_dygraph_Compatibility(self):
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py:260]: 这里按照测试规范来填条目吧，不用自由发挥：

https://github.com/PaddlePaddle/docs/blob/develop/docs/dev_guides/coding_agent/api_compatibility/python-decorator.md#step-5-%E6%B7%BB%E5%8A%A0%E6%B5%8B%E8%AF%95%E7%94%A8%E4%BE%8B

File: test/legacy_test/test_aminmax_op.py [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+            )
+
+            ref_min = np.amin(self.np_x)
+            ref_max = np.amax(self.np_x)
+            for i in range(0, len(fetches), 2):
+                np.testing.assert_allclose(fetches[i], ref_min)
+                np.testing.assert_allclose(fetches[i + 1], ref_max)
+
+
+if __name__ == '__main__':
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py:438]: 还需要加一下 `paddle.jit.to_static(backend='CINN')` 的测试，覆盖AminmaxOpInferSymbolicShape的测试

File: test/legacy_test/test_aminmax_op.py (hunk 14 above)
Comment (Manfredss) [reply test/legacy_test/test_aminmax_op.py:118]: float32 精度下（约7位有效数字），0.005 的扰动会改变 min/max 落在哪个元素上，导致数值梯度和解析梯度完全不一致，最大误差达到 1.0。float64  精度足够，元素间的数值差距远大于扰动量，不会出现这个问题

File: test/legacy_test/test_aminmax_op.py [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+
+class TestAminmaxDynamicShape(unittest.TestCase):
+    def setUp(self):
+        np.random.seed(2025)
+        self.shape = [4, 5, 6]
+        self.dtype = "float32"
+        self.init_shape = [None, None, 6]
+        self.x = np.random.random(self.shape).astype(self.dtype)
+        self.net = aminmax_net
+        self.enable_cinn = False
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py]: 你这没开cinn，导致还是没覆盖到

File: test/legacy_test/test_aminmax_op.py [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+            ref_max = np.amax(self.np_input)
+            np.testing.assert_allclose(fetches[-2], ref_min)
+            np.testing.assert_allclose(fetches[-1], ref_max)
+
+
+def aminmax_net(x):
+    return paddle.aminmax(x, axis=1)
+
+
+def apply_to_static(net, use_cinn, input_spec=None):
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py]: 这一层建议不包了，包了太多层了

File: test/legacy_test/test_aminmax_op.py [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+        self.dtype = "float32"
+        self.init_shape = [None, None, 6]
+        self.x = np.random.random(self.shape).astype(self.dtype)
+        self.net = aminmax_net
+        self.enable_cinn = False
+        self.tol = 1e-6
+
+    def base_net(self, flag=None):
+        x = paddle.to_tensor(self.x)
+        if flag == "static":
```
Comment (zhwesky2010) [inline test/legacy_test/test_aminmax_op.py:380]: if "static"

elif "dygraph"

**Review by zhwesky2010 — COMMENTED** [review]
test_api_compatibility_part3.py的TestConv2dTransposeAPI有点问题，导致现在CI全部无法通过了，这个PR注释下case或者修复

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

File: python/paddle/tensor/__init__.py [hunk 20]
```diff
@@ -622,9 +622,9 @@
     'max',
     'amax',
     'maximum',
-    'positive',
     'min',
     'amin',
+    'aminmax',
```
Comment (zhwesky2010) [inline python/paddle/tensor/__init__.py:628]: example code挂了，这个没用移除下 

CI可能是用develop的版本跑的示例代码，示例代码下个PR再加吧

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/unary_infer_sym.cc [hunk 21]
```diff
@@ -307,6 +307,24 @@ bool AminOpInferSymbolicShape(pir::Operation *op,
                                  axis.size() == 0 /*reduce_all*/);
 }
 
+bool AminmaxOpInferSymbolicShape(
+    pir::Operation *op, pir::InferSymbolicShapeContext *infer_context) {
+  const auto &axis = details::GetVectorAttr(op, "axis");
```
Comment (zhwesky2010) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/unary_infer_sym.cc:312]: 这个打印个信息调试下吧，显示没覆盖到

**Comment by Manfredss** [comment]
屏幕截图 2026-04-12 081133


@zhwesky2010  加了打印信息在跑 pyhton test/legacy_test/test_aminmax_op.py 的时候会有打印，这是覆盖到了吧？

**Comment by zhwesky2010** [comment]
@Manfredss 看起来是已经正常覆盖了

**Review by zhwesky2010 — APPROVED** [review]
LGTM

**Comment by zhwesky2010** [comment]
@Manfredss 示例代码还需要先关一下

File: python/paddle/_paddle_docs.py [hunk 22]
```diff
@@ -320,6 +320,82 @@ def amin(
...
+
+            >>> min_val, max_val = paddle.aminmax(x, axis=0)
+            >>> min_val
+            Tensor(shape=[4], dtype=float64, place=Place(cpu), stop_gradient=False,
+            [0.10000000, 0.90000000, 0.60000000, 0.70000000])
+            >>> max_val
+            Tensor(shape=[4], dtype=float64, place=Place(cpu), stop_gradient=False,
+            [0.90000000, 0.90000000, 0.90000000, 0.90000000])
+
+            >>> min_val, max_val = paddle.aminmax(x, axis=1, keepdim=True)
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py]: 这些示例代码跳过失效了，这个可能只是跳过对输出结果的检查。不跳过示例代码本身的执行？

**Comment by zhwesky2010** [comment]
需要看一下CI结果，这个不是随机挂，rerun只针对随机挂才有用
infoflow 2026-04-23 15-15-36

**Review by zhwesky2010 — APPROVED** [review]
LGTM

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by SigureMo — APPROVED** [review]
